src/core/document.js

```diff
@@ -715,6 +697,61 @@ class PDFDocument {
     return shadow(this, "numPages", num);
   }
 
+  /**
+   * @private
+   */
+  _hasOnlyDocumentSignatures(fields) {
+    return fields.every(field => {
+      field = this.xref.fetchIfRef(field);
+      if (field.has("Kids")) {
+        return this._hasOnlyDocumentSignatures(field.get("Kids"));
+      }
+      const fieldType = field.get("FT");
+      const isSignature = isName(fieldType) && fieldType.name === "Sig";
+      const rectangle = field.get("Rect");
+      const isInvisible =
+        Array.isArray(rectangle) && rectangle.every(value => value === 0);
+      return isSignature && isInvisible;
+    });
+  }
+
+  get formInfo() {
+    const formInfo = { hasAcroForm: false, hasXfa: false };
+    try {
+      const acroForm = this.catalog.acroForm;
+      if (acroForm) {
+        // The document contains XFA data if the `XFA` entry is a non-empty
+        // array or stream.
+        const xfa = acroForm.get("XFA");
+        const hasXfa = (Array.isArray(xfa) || isStream(xfa)) && xfa.length > 0;
+        formInfo.hasXfa = hasXfa;
+
+        // The document contains AcroForm data if the `Fields` entry is a
+        // non-empty array and it doesn't consist of only document signatures.
+        // This second check is required for files that don't actually contain
+        // AcroForm data (only XFA data), but that use the `Fields` entry to
+        // store (invisible) document signatures. This can be detected using
+        // the first bit of the `SigFlags` integer (see Table 219 in the
+        // specification).
+        const fields = acroForm.get("Fields");
+        const hasFields = Array.isArray(fields) && fields.length > 0;
+        const sigFlags = acroForm.get("SigFlags");
+        const hasDocumentSignatures = !!(sigFlags & 0x1);
+        let hasOnlyDocumentSignatures = false;
+        if (hasDocumentSignatures) {
+          hasOnlyDocumentSignatures = this._hasOnlyDocumentSignatures(fields);
+        }
```

Comment:
Nit: You can probably collapse this a little bit, since there's intermediate state that's not really being used here.
```suggestion
        const hasOnlyDocumentSignatures =
          !!(sigFlags & 0x1) && this._hasOnlyDocumentSignatures(fields);
          
```